src/Aspire.Dashboard/Telemetry/DashboardTelemetrySender.cs

```diff
@@ -0,0 +1,202 @@
+    {
+        _options = options;
+        _logger = logger;
+        var channelOptions = new BoundedChannelOptions(1000)
+        {
+            FullMode = BoundedChannelFullMode.DropOldest,
+            SingleReader = true
+        };
+        _channel = Channel.CreateBounded<(OperationContext, Func<HttpClient, Func<OperationContextProperty, object>, Task>)>(channelOptions);
+    }
+
+    private static bool HasDebugSession(
+        DebugSession debugSession,
+        [NotNullWhen(true)] out Uri? debugSessionUri,
+        [NotNullWhen(true)] out string? token,
+        [NotNullWhen(true)] out X509Certificate2? serverCert)
+    {
+        if (debugSession.Address is not null && debugSession.Token is not null && debugSession.GetServerCertificate() is { } cert && debugSession.TelemetryOptOut is not true)
+        {
+            debugSessionUri = new Uri($"https://{debugSession.Address}");
+            token = debugSession.Token;
+            serverCert = cert;
+            return true;
+        }
+
+        debugSessionUri = null;
+        token = null;
+        serverCert = null;
+        return false;
+    }
+
+    private void StartSendLoop()
+    {
+        Debug.Assert(_httpClient is not null, "HttpClient must be initialized.");
+
+        _sendLoopTask = Task.Run(async () =>
+        {
+            _logger.LogInformation("Starting sender loop.");
+
+            try
+            {
+                while (await _channel.Reader.WaitToReadAsync().ConfigureAwait(false))
+                {
+                    while (_channel.Reader.TryRead(out var operation))
+                    {
+                        var (context, requestFunc) = operation;
+                        try
+                        {
+                            await requestFunc(_httpClient, GetResponseProperty).ConfigureAwait(false);
+
+                            // Double check properties are set.
+                            foreach (var property in context.Properties)
+                            {
+                                if (!property.HasValue)
+                                {
+                                    _logger.LogWarning("Unset context property.");
+                                }
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            _logger.LogWarning(ex, "Failed to send telemetry request.");
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                _logger.LogInformation("Ending sender loop.");
+            }
+        });
+    }
+
+    private HttpClient CreateHttpClient(Uri debugSessionUri, string token, X509Certificate2 cert)
+    {
+        var handler = new HttpClientHandler
+        {
+            ServerCertificateCustomValidationCallback = (_, c, _, e) =>
```

Comment:
Only set this callback if a cert is provided. Otherwise requests will always be http